streamer/src/quic.rs

```diff
@@ -0,0 +1,188 @@
+    // We have to convert it to DER and pass it to the library.
+
+    // Convert private key into PKCS#8 v1 object.
+    // RFC 8410, Section 7: Private Key Format
+    // https://datatracker.ietf.org/doc/html/rfc8410#section-7
+    let mut private_key = Vec::<u8>::with_capacity(34);
+    private_key.extend_from_slice(&[0x04, 0x20]); // ASN.1 OCTET STRING
+    private_key.extend_from_slice(identity_keypair.secret().as_bytes());
+    let key_pkcs8 = pkcs8::PrivateKeyInfo {
+        algorithm: AlgorithmIdentifier {
+            oid: ObjectIdentifier::from_arcs(&[1, 3, 101, 112]).unwrap(),
+            parameters: None,
+        },
+        private_key: &private_key,
+        public_key: None,
+    };
+    let key_pkcs8_der = key_pkcs8
+        .to_der()
+        .expect("Failed to convert keypair to DER")
+        .to_der();
+
+    // Parse private key into rcgen::KeyPair struct.
+    let keypair =
+        rcgen::KeyPair::from_der(&key_pkcs8_der).expect("Failed to parse keypair from DER");
+
+    let mut cert_params = CertificateParams::default();
+    cert_params.subject_alt_names = vec![SanType::IpAddress(san)];
+    cert_params.alg = &rcgen::PKCS_ED25519;
+    cert_params.key_pair = Some(keypair);
+    cert_params.distinguished_name = DistinguishedName::new();
+    cert_params
+        .distinguished_name
+        .push(DnType::CommonName, "Solana node");
+    cert_params
+}
+
+fn rt() -> Runtime {
+    Builder::new_current_thread().enable_all().build().unwrap()
+}
+
+pub fn spawn_server(
+    sock: UdpSocket,
+    keypair: &Keypair,
+    gossip_host: IpAddr,
+    packet_sender: Sender<PacketBatch>,
+    exit: Arc<AtomicBool>,
+) -> Result<thread::JoinHandle<()>, Box<dyn Error>> {
+    let (config, _cert) = configure_server(keypair, gossip_host)?;
+    let handle = thread::spawn(move || {
+        let runtime = rt();
+        let (_, mut incoming) = {
+            let _guard = runtime.enter();
+            Endpoint::new(EndpointConfig::default(), Some(config), sock).unwrap()
+        };
+        let handle = runtime.spawn(async move {
+            while !exit.load(Ordering::Relaxed) {
+                let timeout_connection =
+                    timeout(Duration::from_millis(1000), incoming.next()).await;
+
+                if let Ok(connection) = timeout_connection {
+                    let quinn::NewConnection {
+                        connection,
+                        mut uni_streams,
+                        ..
+                    } = connection.unwrap().await.expect("connect");
+
+                    let remote_addr = connection.remote_address();
+                    while let Some(Ok(mut stream)) = uni_streams.next().await {
+                        let packet_sender = packet_sender.clone();
+                        tokio::spawn(async move {
+                            while let Some(chunk) =
+                                stream.read_chunk(usize::MAX, false).await.unwrap()
```

Comment:
is `usize::MAX` here what we want or something that's on the order of a packet?

Reply:
This `unwrap()` on the `read_chunk()` also doesn't seem safe. I think things would be a little clearer about the error types/await results if we restructured the top level `while let Some(Ok(mut stream)) = uni_streams.next().await {` loop with some annotations like this:

```
while let Some(Ok(mut recv_stream)) = {
    let recv_stream: Option<Result<quinn::RecvStream, quinn::ConnectionError>> =
        uni_streams.next().await;
    recv_stream
} {
    let packet_sender = packet_sender.clone();
    tokio::spawn(async move {
        while let Ok(Some(chunk)) = {
            let chunk: Result<Option<quinn::Chunk>, quinn::ReadError> =
                recv_stream.read_chunk(usize::MAX, false).await;
            chunk
        } {
            trace!("got chunk: {:?}", chunk);
            let mut batch = PacketBatch::with_capacity(1);
            let mut packet = Packet::default();
            let len = chunk.bytes.len();
            packet.meta.set_addr(&remote_addr);
            packet.meta.size = len;
            packet.data[..len].copy_from_slice(&chunk.bytes);
            batch.packets.push(packet);
            if let Err(e) = packet_sender.send(batch) {
                info!("send error: {}", e);
            } else {
                trace!("sent {} byte packet", len);
            }
        }
    });
}
```